KAFKA-3396 : Unauthorized topics are returned to the user

edoardocomar opened the pull request:
Reopening of https://github.com/apache/kafka/pull/1428

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

Modified KafkaApis to return Errors.UNKNOWN_TOPIC_OR_PARTITION if
principal has no Describe access to topic

Unit tests expanded

Some paths cause the client to block due to bug
https://issues.apache.org/jira/browse/KAFKA-3727?filter=-2
tests work around this by executing in separate thread

mimaison committed:
KAFKA-3396: Updates and cleanups following the feedback

mimaison committed:
KAFKA-3396: More updates + rebased off trunk

- Added tests from 44ad3ec
- Small refactorings

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

Rebased after kip-79 changes.
Fixing leak of topic for LIST_OFFSETS when unauthorized.
Added tests.

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

cleanup after review

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

Cleanup addressing comments

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

Cleanup addressing comments

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

Revised handling of FETCH PRODUCE and DELETE requests

ijuma commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -247,22 +260,22 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
+//      ApiKeys.LEADER_AND_ISR -> createLeaderAndIsrRequest,
+//      ApiKeys.STOP_REPLICA -> createStopReplicaRequest,
+//      ApiKeys.CONTROLLED_SHUTDOWN_KEY -> createControlledShutdownRequest,
+//      ApiKeys.CREATE_TOPICS -> createTopicsRequest,
```
This seems unintentional?

edoardocomar replied:
oops

edoardocomar replied:
thanks for spotting that shameful oops, @ijuma . full test passes

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

oops had commented out part os a test by mistake

edoardocomar commented:
Hi @ijuma all tests seem to have passed (they do locally for me) but the Jenkins build failed

ijuma commented:
Thanks @edoardocomar, will take another look soon. Can you please add a note to `upgrade.html`?

ijuma commented on core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:
```diff
@@ -149,18 +173,28 @@ trait EndToEndAuthorizationTest extends IntegrationTestHarness with SaslSetup {
     TestUtils.createTopic(zkUtils, topic, 1, 3, this.servers)
   }
 
+  override def createNewProducer() : KafkaProducer[Array[Byte], Array[Byte]] = {
```
There should be no space before the `:`. There are a few other cases like that.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1176,21 +1198,17 @@ class KafkaApis(val requestChannel: RequestChannel,
-      if (authorize(request.session, Describe, new Resource(auth.Topic, topic))) {
-        (topic, Errors.TOPIC_AUTHORIZATION_FAILED)
-      } else {
-        (topic, Errors.INVALID_TOPIC_EXCEPTION)
```
I think the usage of `INVALID_TOPIC_EXCEPTION` is intentional here. One of the differences is that `INVALID_TOPIC_EXCEPTION` is not retriable. @granthenke, is that right?

granthenke replied:
I think changing to a UNKNOWN_TOPIC_OR_PARTITION exception might make sense in this case. Especially if that is the behavior we are using for the other requests where a topic is not authorized and we don't want to leak the information. Consistency is always easier to reason about. 

The most important thing is that the exception doesn't give any indication of whether or not the topic exists. Its behavior should the same as if the topic isn't there at all. If this exception is changed, I think the InvalidTopicException in AdminUtils may need to be changed too:
https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/admin/AdminUtils.scala#L328

edoardocomar replied:
done

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

addressing comments: code formatting, not in upgrade.html

edoardocomar commented:
added a note to `upgrade.html` in the `upgrade_1010_notable` section

ijuma commented on docs/upgrade.html:
```diff
@@ -62,6 +62,9 @@
     <li> The BrokerState "RunningAsController" (value 4) has been removed. Due to a bug, a broker would only be in this state briefly before transitioning out of it and hence the impact of the removal should be minimal. The recommended way to detect if a given broker is the controller is via the kafka.controller:type=KafkaController,name=ActiveControllerCount metric. </li>
     <li> The new Java Consumer now allows users to search offsets by timestamp on partitions.
+    <li> When using an Authorizer and a user hasn't got <b>Describe</b> authorization on a topic, the broker will no longer return TOPIC_AUTHORIZATION_FAILED errors  
+    but just UNKNOWN_TOPIC_OR_PARTITION errors, to avoid leaking topic names.
```
Closing `<\li>` missing

edoardocomar replied:
done

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

addressing comments: missing </li>

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -234,44 +234,44 @@ class KafkaApis(val requestChannel: RequestChannel,
           if (errorCode != Errors.NONE.code) {
-            debug(s"Offset commit request with correlation id ${header.correlationId} from client ${header.clientId} " +
-              s"on partition $topicPartition failed due to ${Errors.forCode(errorCode).exceptionName}")
+            if (errorCode == Errors.UNKNOWN_TOPIC_OR_PARTITION.code && metadataCache.contains(topicPartition.topic)) 
```
Would it be easier to handle the logging for this case in the initial call to `partition` above? May as well save the redundant cache lookup and it seems possible that the cache may have changed in between the first check and this one. Another minor point: is it worthwhile checking if DEBUG is enabled prior to doing this loop?

edoardocomar replied:
done thanks

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -337,15 +337,22 @@ class KafkaApis(val requestChannel: RequestChannel,
     val numBytesAppended = request.header.sizeOf + produceRequest.sizeOf
 
-    val (authorizedRequestInfo, unauthorizedRequestInfo) = produceRequest.partitionRecords.asScala.partition {
+    val (existingAndAuthorizedForDescribeTopics, nonExistingOrUnauthorizedForDescribeTopics) = produceRequest.partitionRecords.asScala.partition {
```
For discussion: Since `Write` access subsumes `Describe`, I'm wondering if it would make sense to change the order of the checks to something like this:
1. Check for Write access on all topics.
2. For those topics without Write access:
   1. If no Describe access or not exists -> UNKNOWN_TOPIC_OR_PARTITION
   2. Otherwise -> TOPIC_AUTHORIZATION_FAILED

The intuition is that failing `Write` should be the exceptional case. The one annoying thing is that you also need to check existence for topics which are authorized for `Write`. The fetch request handler could be done similarly since `Read` access also subsumes `Describe`.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -811,25 +825,34 @@ class KafkaApis(val requestChannel: RequestChannel,
-    val unauthorizedTopicMetadata = unauthorizedTopics.map(topic =>
-      new MetadataResponse.TopicMetadata(Errors.TOPIC_AUTHORIZATION_FAILED, topic, Topic.isInternal(topic),
+    val unauthorizedForCreateTopicMetadata = unauthorizedForCreateTopics.map(topic =>
+      new MetadataResponse.TopicMetadata(Errors.TOPIC_AUTHORIZATION_FAILED, topic, common.Topic.isInternal(topic),
```
@ijuma Is it correct that we return TOPIC_AUTHORIZATION_FAILED here? Whether or not it's ideal, the authorization check is on the `Cluster` resource currently.

ijuma replied:
That's a bit weird, isn't it? In a KIP discussion, we had said that we wanted to change it so that one would require a `Create Topic` instead of `Create Cluster` ACL for topic creation (would probably need both for compatibility), so it seems reasonable to keep throwing a `TOPIC_AUTHORIZATION_FAILED` here.

hachikuji replied:
Ack. I agree that this is probably better long term, even if it's a little misleading at the moment.

hachikuji reviewed (commented):
Left a couple more comments. It's looking pretty good on the server side. However, I still do not see any checks on the client side in the offset fetch and offset commit response handlers for the UNKNOWN_TOPIC_OR_PARTITION error code. I feel it would also be helpful when we catch this error explicitly and retry internally to log a message which mentions that the user may not have Describe access. Is there any reason we shouldn't do this?

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

addressing comments: 
handling UTOP in ConsumerCoordinator offset fetch/commit
improved logging efficiency on the server

edoardocomar commented:
Handling UTOP in consumerCoordinator handleResponse for fetchOffset and commitOffset, with more descriptive msg

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

test cleanup and fixing intermittent failure

edoardocomar commented:
rebased/merged and fixed spurious test failure

edoardocomar commented:
@ijuma the Jenkins build failed to complete again. However all tests pass locally for me

edoardocomar committed:
KAFKA-3396 : Unauthorized topics are returned to the user

throwing UTOP also form AdminUtils.delete

hachikuji commented:
I've been running system tests against this branch. There were a couple test failures, but I think they are known transient failures. I'm running again to be sure. In the meantime, I'll do another pass over the code and hopefully we can merge this today.

edoardocomar commented:
@hachikuji I rebased and rerun all the tests locally on my mac and they pass.

hachikuji commented:
LGTM. Merging to trunk and 0.10.1. Thanks @edoardocomar and @mimaison for the hard work (and patience) on this patch! I have some minor cleanups/improvements on the client and in testing, which I will submit in a follow-up PR.

hachikuji commented:
By the way, I changed the title. Hopefully it's reasonable.
